resource/graph_node.go

```diff
@@ -0,0 +1,309 @@
+package resource
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+	"go.viam.com/utils"
+)
+
+// A GraphNode contains the current state of a resource.
+// It starts out as either uninitialized, unconfigured, or configured.
+// Based on these states, the underlying Resource may or may not be available.
+// Additionally, the node can be informed that the resource either needs to be
+// updated or eventually removed. During its life, errors may be set on the
+// node to indicate that the resource is no longer available to external users.
+type GraphNode struct {
+	mu                        sync.RWMutex
```

Comment:
Same question here about how we know which members the mutex is protecting access to.

Reply:
it's all within the same file in exported methods which is usually the common pattern!